src/vcpkg-test/tools.cpp

```diff
@@ -1,43 +1,69 @@
 #include <catch2/catch.hpp>
 
 #include <vcpkg/tools.h>
-
-#include <array>
+#include <vcpkg/versions.h>
 
 using namespace vcpkg;
 
+TEST_CASE ("tool_version_from_values", "[tools]")
+{
+    auto result = ToolVersion::from_values({});
+    CHECK(result.original_text == "");
+    CHECK(result.version == std::vector<uint64_t>{});
+
+    result = ToolVersion::from_values({42});
+    CHECK(result.original_text == "42");
+    CHECK(result.version == std::vector<uint64_t>{42});
+
+    result = ToolVersion::from_values({42, 1729});
+    CHECK(result.original_text == "42.1729");
+    CHECK(result.version == std::vector<uint64_t>{42, 1729});
+}
+
 TEST_CASE ("parse_tool_version_string", "[tools]")
 {
-    auto result = parse_tool_version_string("1.2.3");
+    auto result = ToolVersion::try_parse_numeric("1.2.3");
     REQUIRE(result.has_value());
-    CHECK(*result.get() == std::array<int, 3>{1, 2, 3});
+    CHECK(*result.get() == ToolVersion::from_values({1, 2, 3}));
 
-    result = parse_tool_version_string("3.22.3");
+    result = ToolVersion::try_parse_numeric("3.22.3");
     REQUIRE(result.has_value());
-    CHECK(*result.get() == std::array<int, 3>{3, 22, 3});
+    CHECK(*result.get() == ToolVersion::from_values({3, 22, 3}));
 
-    result = parse_tool_version_string("4.65");
+    result = ToolVersion::try_parse_numeric("4.65");
     REQUIRE(result.has_value());
-    CHECK(*result.get() == std::array<int, 3>{4, 65, 0});
+    CHECK(*result.get() == ToolVersion::from_values({4, 65, 0}));
 
-    result = parse_tool_version_string(R"(cmake version 3.22.2
+    result = ToolVersion::try_parse_numeric(R"(cmake version 3.22.2
 CMake suite maintained and supported by Kitware (kitware.com/cmake).)");
     REQUIRE(result.has_value());
-    CHECK(*result.get() == std::array<int, 3>{3, 22, 2});
+    CHECK(*result.get() == ToolVersion::from_values({3, 22, 2}));
 
-    result = parse_tool_version_string(R"(aria2 version 1.35.0
+    result = ToolVersion::try_parse_numeric(R"(aria2 version 1.35.0
 Copyright (C) 2006, 2019 Tatsuhiro Tsujikawa)");
     REQUIRE(result.has_value());
-    CHECK(*result.get() == std::array<int, 3>{1, 35, 0});
+    CHECK(*result.get() == ToolVersion::from_values({1, 35, 0}));
 
-    result = parse_tool_version_string(R"(git version 2.17.1.windows.2)");
+    result = ToolVersion::try_parse_numeric(R"(git version 2.17.1.windows.2)");
     REQUIRE(result.has_value());
-    CHECK(*result.get() == std::array<int, 3>{2, 17, 1});
+    CHECK(*result.get() == ToolVersion::from_values({2, 17, 1}));
 
-    result = parse_tool_version_string(R"(git version 2.17.windows.2)");
+    result = ToolVersion::try_parse_numeric(R"(git version 2.17.windows.2)");
     REQUIRE(result.has_value());
-    CHECK(*result.get() == std::array<int, 3>{2, 17, 0});
+    CHECK(*result.get() == ToolVersion::from_values({2, 17, 0}));
 
-    result = parse_tool_version_string("4");
+    result = ToolVersion::try_parse_numeric("4");
```

Comment:
I'd also like a test case of `"1.2.5."` (trailing dot). This could be created by the `try_parse_git()` algorithm given strange input.

Reply:
Done.